### ISSUE TITLE
The big "I want to test multiple pages/upload images/open popups" issue

### ISSUE THREAD

**Issue by soulgalore** [sitespeedio/sitespeed.io#1827]
Very often we have issues where you as a user want to test a user flow: You want to access one URL, add something to the cart, then go to a new page etc. And then have measurements for everything. Or you maybe want to upload something and measure that. Or you want to click a link and measure it.

What we have today:
Today we have support for running a Selenium NodeJS script before you go to a URL (used to log in the user or preload the browser cache). You can also run a postScript that is a Selenium script that run after the URL is tested.

For 6.0 I thought we could have something that could support clicking a link but we didn't have time to do it. @tobli and me has discussed and we feel that we unsure how it would work and if it really would work (would the metrics really be stable?).

However we are open for discussion, PRs and proof of concepts! Talk to us here, send a PR how you can do it or do a POC.

We will redirect all related issues to this issue.

**Comment by Nuoky** [comment]
I implemented some DOM operations through the Chrome plugin, but I can only solve this problem by using the pageCompleteCheck to set the test time.

Sorry, google translate.

* example：
```
--browsertime.chrome.args load-extension="./chromeplugin/h5click" 
```
```
--browsertime.pageCompleteCheck 'return (function() {try { return (Date.now() - window.performance.timing.loadEventEnd) > 10000;} catch(e) {} return true;})()'
```

**Comment by soulgalore** [comment]
@Nuoky thanks for sharing!

**Comment by Algram** [comment]
Is there any chance of doing this with the current tooling already or does this need changes in browsertime?

**Comment by soulgalore** [comment]
@Algram what exactly do you need to do? some parts can be done, some parts not, and some parts we believe is not possible to get stable metrics and there's probably better ways to do it.

**Comment by gijswijs** [comment]
I've been thinking about extending sitespeed.io with synthetic monitoring, and this is what I came up with. In it's core it should be one selenium script that should be executed.

## Measuring performance
Use a miniprofiler kind of tool (https://miniprofiler.com/) to measure the execution steps of the selenium script
This would result in a browsertime.json kind of file that holds the timings of the entire script and each separate step.
In grafana this could be visualized using a stacked line chart.
(source: https://www.pingdom.com/lp/transaction-monitoring)

## Three types of errors
 - Availability errors: Availability errors occur when the application does not respond, or when the application responds with an error that indicates it is unavailable.
 - Accuracy errors: Accuracy errors occur when an assertion made in the selenium script fails. Functions that can be evaluated to determine accuracy include link checking, content validation, title validation, and response data verification.
 - Script errors: An execution error indicates that there were problems with the execution of the selenium script.

 This results in three new metrics, namely the number of each type of error occuring.

(source: https://docs.bmc.com/docs/display/tsavm107/Synthetic+metrics%2C+rules%2C+and+events)

## Video
Browsertime should start FFMPEG to record a video of the entire script execution.

## Should the separate pages/urls have a full sitespeed.io report as well?

It's debatable whether you should have separate performance reports for the pages the script traverses. A practical issue is: what is a page? Does any changed uri result in a different page? I would think not. So maybe you would have to configure the separate page reports manually. You couldn't run the full page reports along the synthetic monitoring, because the page reporting affects the performance of the script. A script could wait until a certain object in the dom is available, and then carry on. On the other hand a page report would have to wait until the entire dom is loaded to be able to do a full report. My stance is that page performance should be separate from synthetic performance monitoring.

So the fit with sitespeed isn't perfect, but I think sitespeed *is* the perfect foundation to build synthetic monitoring on top of.

What do you guys think?

**Comment by soulgalore** [comment]
Hi @gijswijs thanks for spending time on thinking on this, let us digest and think about it during the Christmas break. 

One thing I've been thinking about is to make it possible to run multiple URLs in Browsertime (as a first step). You could add X URLs and it would test them all one by one with no clearing the cache in between (that would match the "user journey" thing we've been talking about the last years).

What me and @tobli has been skeptical about (don't know your thoughts @beenanner) is how accurate the metrics will be and what kind of regressions you will be able to find using that kind of tests, where you test multiple pages, scripts, and do one run each? For example running WebPageTest on c4.large doing one access (old Windows version), testing one URL 5 times for Wikipedia makes it possible to find a regression of 15-20%. If we then test multiple pages with just one run, I think the percentage will be quite high? However I haven't spent anytime testing it out so if someone got some experience and can share it, it would be great! Maybe there are closed source tools out there that works? I've never tried but would be really interesting if someone got something to share :)

When we do the tests with WebPageReplay (the new container we released sometime ago), we can see that we can spot regressions with Browsertime that is 1-2% and that's more of where we wanna be.

Best
Peter

**Comment by gijswijs** [comment]
Hi @soulgalore,

I'm not sure whether I understand your concept of regression in this particular context.
I'm guessing you are talking about performance regression? And you want to be able to find the smallest, but still significant and relevant, change in performance?

In that case, I think you're right, it would be hard to spot regressions of 1-2%. You could probably find smaller regressions (but not *that* small) when you compare multiple tests combined over a period of time (a day) with a previous period (yesterday), but that's not primarily what I'm interested in.

The way we use synthetic monitoring is as an extension to up-time monitoring. So a website is either up or down, and the particulars of the actual performance are of lesser importance. But literally pinging the website and seeing if you get response code 200 back just doesn't cut it. We want to see whether it's possible to perform certain tasks, like searching for a product, adding it to your cart, and buying it.
That's a bit more complex than running multiple URLs without clearing the cache in between. We need to be able to search for specific DOM elements, post forms, etc. That's why Selenium is such a good fit.

When running those selenium tests I'm primarily interested in whether they run at all, and if not, what's causing them to fail. Hence the three different types of errors. The performance, while still very important, is secondary. I would definitely be OK with not being able to find a regression of less than 10-15%

Best regards,
Gijs

**Comment by donwalling** [comment]
@soulgalore  has there been any additional thinking on this topic. We've just picked up sitespeed a few days ago, and made amazing progress using the pageCompleteCheck strategy mentioned above. But this will soon become a maintenance burden as we scale. We'd like to keep pageCompleteCheck lightweight as the simple timer that it is today, and use a different for driving the user flow. 

What would meet our needs is the ability to specify a user flow script, ideally a selenium script, that would be executed a configurable number of milliseconds after loadEventEnd. Is that along the lines of what you're thinking for this improvement?

Thanks,
Don

**Comment by soulgalore** [comment]
Hey @donwalling not sure I follow 100%. Can you show me step by step how you mean? 

Best
Peter

**Comment by soulgalore** [comment]
I've been rethinking this and started to try out a new implementation. Easiest is to do it by two steps. The first step would make it possible for people to measure one page (as today) but where you can run whatever selenium script you want (meaning you can test login to a page, add to your cart, clicking a popu and then move etc). This will be non breaking change and we could release it in the next release where we add new functionality.

Testing multiple pages would be going out in the next major. Maybe we can do it as non breaking but I need to implement it first to be sure.

**Comment by donwalling** [comment]
We're happy to help test. The first step described is what we're looking for. I haven't figured out how to implement, but will try to help in any way we can

**Comment by soulgalore** [comment]
@donwalling cool, I'll share the branch in this issue when I push it, I wanna discuss it with @tobli and @beenanner first.

**Comment by soulgalore** [comment]
@donwalling We got a branch now in Browsertime that you can try out if you have time: https://github.com/sitespeedio/browsertime/tree/navigationScript#script-navigation

**Comment by donwalling** [comment]
@soulgalore great, thanks for the update. I should be able to take a look within the next few days.

**Comment by donwalling** [comment]
@soulgalore  I just read through the docs (I haven't actually tried to do anything yet, so I apologize if I haven't fully understood the approach). How is the script specified to browsertime - is it still a preScript? For our use case, we'd want something very similar to the "Testing a page after you have logged in:" case where after login we'd make a call similar to help.startAndNavigate(context.url)   except that it'd be something like help.startAndRunSeleniumScript(context.testScript) - i.e. after login we'd wait for something on the page that verifies that we are logged in, and then begin some SPA interactions (not a navigation). I'll see how far i can get with what you've done, and let you know how it goes.

**Comment by soulgalore** [comment]
Oops, sorry I missed adding that to the docs (only in the cli help). Use `--scriptNavigation`. That is a Selenium script following the same structure as before + an extra helper object for starting to measure/navigate. In your case it would then be:

```
module.exports = {
  run(context, help) {
        await help.navigate('URL');
        await for that special thing in the page
        await help.start() // now we start to measure, if video is enabled the video will be started
        await your interactions
    });
  }
};
```

Maybe that will work :)

**Comment by donwalling** [comment]
Hi @soulgalore,  a colleague @Siwach16 has made great progress moving a couple of our tests to use this approach. It's working well. We have no issues or concerns to report so far, and are planning on adopting it fully when released.

One thing we noticed on the pull request https://github.com/sitespeedio/browsertime/pull/666 is a comment that preScript/postScript should be deprecated. We're planning on continuing to use preScript for login that is common across all of our tests, since it provides a good modularity approach in our script files - pre/post script can be common things across all tests, and navigationScript is the specific test script. We'd prefer pre/post scripts not be deprecated for this reason.

**Comment by soulgalore** [comment]
Ok cool, lets keep it like that then! Thanks for the feedback!

**Comment by soulgalore** [comment]
Hey @PieterVO do you mean sitespeedio/browsertime#666 ? It will come in next major of sitespeed.io. If wanna try it out in Docker with Browsertime you can try out our autobuild: sitespeedio/browsertime-autobuild:navigationScript.

The schedule will look like this: Coach 3.0 (maybe this week), Browsertime 4 and then sitespeed.io 8. I plan to release sitespeed.io before Christmas but exact date is hard to say, it depends on if testing works out ok.

**Comment by PieterVO** [comment]
Thanks for your fast reply @soulgalore. Do I get the full sitespeed.io image when i use _sitespeedio/browsertime-autobuild:navigationScript_?

**Comment by soulgalore** [comment]
No only Browsertime, so either you need build it yourself or wait until we merge and release Browsertime, sorry.

**Comment by Siwach16** [comment]
@soulgalore Can this(cli) change be included as an upcoming patch in the current sitespeed version? This is because the required version of browsertime is already bundled and we just need a way to invoke it using the "scriptNavigation" parameter. We would then not need to wait for the 8.0 release.
Also, w.r.t to ur reply above, I am not quite clear about what test runs of "Travis config files" mean?

**Comment by donwalling** [comment]
Hi @soulgalore  I'm not following how the multi.js referenced at https://github.com/sitespeedio/sitespeed.io/blob/master/.travis.yml#L37 is equivalent to the scriptNavigation option that was previously here https://github.com/sitespeedio/browsertime/pull/666 ? Will --scriptNavigation be re-added at some point? it looked very promising for our use case.
Thanks,
Don

**Comment by soulgalore** [comment]
As a Browsertime user you can run a script from the command line or give It a URL, so instead of adding a extra parameter like `--navigationScript script.js` you just pass on `script.js` (or multiple scripts or scripts and then a URL if you want that). But if you just give me some time to add documentation and do a stable release, it will be easier to understand how to use it :)

**Comment by soulgalore** [comment]
Released 8.0.0 today that should solve most of these use cases, please try it out and read the docs: https://www.sitespeed.io/documentation/sitespeed.io/scripting/
